### ISSUE TITLE
Retraining xgboost model with different loss function

### ISSUE THREAD

**Issue by manikanthr5** [dmlc/xgboost#5359]
## Context: 
Let us define 2 loss functions (taken from [this page](https://github.com/dmlc/xgboost/blob/master/demo/guide-python/custom_rmsle.py)):

XGBoost Version: 0.90
Python Version: 3.5.2
Number of epochs: 50

### Loss Function 1 

```python
def native_rmse(dtrain: xgb.DMatrix,
                dtest: xgb.DMatrix, prev_model:xgb.core.Booster=None) -> Dict[str, Dict[str, List[float]]]:
    '''Train using native implementation of Root Mean Squared Loss.'''
    print('Squared Error')
    squared_error = {
        'objective': 'reg:squarederror',
        'eval_metric': 'rmse',
        'tree_method': 'hist',
        'seed': kSeed
    }
    start = time()
    results: Dict[str, Dict[str, List[float]]] = {}
    trained_model = xgb.train(squared_error,
              dtrain=dtrain,
              num_boost_round=kBoostRound,
              xgb_model=prev_model,
              evals=[(dtrain, 'dtrain'), (dtest, 'dtest')],
              evals_result=results)
    print('Finished Squared Error in:', time() - start, '\n')
    return trained_model, results
```

### Loss Function 2
```python
def native_rmsle(dtrain: xgb.DMatrix,
                 dtest: xgb.DMatrix, prev_model:xgb.core.Booster=None) -> Dict[str, Dict[str, List[float]]]:
    '''Train using native implementation of Squared Log Error.'''
    print('Squared Log Error')
    results: Dict[str, Dict[str, List[float]]] = {}
    squared_log_error = {
        'objective': 'reg:squaredlogerror',
        'eval_metric': 'rmsle',
        'tree_method': 'hist',
        'seed': kSeed
    }
    start = time()
    trained_model = xgb.train(squared_log_error,
              dtrain=dtrain,
              num_boost_round=kBoostRound,
              xgb_model=prev_model,
              evals=[(dtrain, 'dtrain'), (dtest, 'dtest')],
              evals_result=results)
    print('Finished Squared Log Error in:', time() - start)
    return trained_model, results
```

**Note**: Let's assume dtrain and dtest are defined.

## Problem:
The procedure I would like to follow to train a XGBoost model is as follows:
1. Train XGBoost Model for 2 epochs using "rmse" loss function and save the model.
```python
rmse_model, rmse_evals = native_rmse(dtrain, dtest)
```

This works and the output is
```
[0]	dtrain-rmse:732.556	dtest-rmse:662.783
[1]	dtrain-rmse:679.064	dtest-rmse:675.687
[2]	dtrain-rmse:627.901	dtest-rmse:702.533
.
.
[48]	dtrain-rmse:66.5061	dtest-rmse:804.446
[49]	dtrain-rmse:62.9484	dtest-rmse:804.296
```

2. Use the trained `rmse_model` and retrain a new model using "rmse" loss functin.
```python
rmse_model1, rmse_evals1 = native_rmse(dtrain, dtest, rmse_model)
```

This also works and gives the following output:
```
[0]	dtrain-rmse:62.3487	dtest-rmse:804.474
[1]	dtrain-rmse:61.0327	dtest-rmse:804.525
.
.
[48]	dtrain-rmse:19.3592	dtest-rmse:808.246
[49]	dtrain-rmse:19.1166	dtest-rmse:808.246
```

3. Use the trained `rmse_model` and train a new model using "rmsle" loss functin.
```python
rmsle_model, rmsle_evals = native_rmsle(dtrain, dtest)
```

**But this doesn't work. It seems to have stuck**. 
```
[0]	dtrain-rmse:62.9484	dtest-rmse:804.296	dtrain-PyRMSLE:10.3872	dtest-PyRMSLE:10.6404
[1]	dtrain-rmse:62.9484	dtest-rmse:804.296	dtrain-PyRMSLE:10.3872	dtest-PyRMSLE:10.6404
.
.
[48]	dtrain-rmse:62.9484	dtest-rmse:804.296	dtrain-PyRMSLE:10.3872	dtest-PyRMSLE:10.6404
[49]	dtrain-rmse:62.9484	dtest-rmse:804.296	dtrain-PyRMSLE:10.3872	dtest-PyRMSLE:10.6404
```

But this works when I trained with 2 epochs (for both loss functions) with following output:
```
[0]	dtrain-rmse:679.078	dtest-rmse:675.713	dtrain-PyRMSLE:3.79744	dtest-PyRMSLE:3.63441
[1]	dtrain-rmse:679.13	dtest-rmse:675.729	dtrain-PyRMSLE:3.79698	dtest-PyRMSLE:3.6231
``` 

It seems that it is not training or reached a local minima. Can we actually do something like this? Train the model for few epochs with one loss function and retrain it with another loss function? Could anybody help me with what is happening in both the cases. 

Thank you.

**Comment by trivialfis** [comment]
> Can we actually do something like this?

Technically yes.  But the objective defines the gradient (and global bias sometimes), so when you change the objective you are training on a different scale (or even different space) of gradient.

**Comment by RAMitchell** [comment]
If you are trying to chain training from different models a good way is to train the first as a separate model, then output predictions as margin, create a new DMatrix on the same data initialising with the margin vector and then train a second model with this DMatrix as input.

**Comment by manikanthr5** [comment]
Ok

> create a new DMatrix on the same data initialising with the margin vector and then train a second model with this DMatrix as input.

How should I do this? Is there any article or issue related to this. Thank you.
